### DISCUSSION TITLE
descriptors for implicit feedback endpoint

### DISCUSSION THREAD

**Discussion by mastupristi** [hathach/tinyusb#2475] [Q&A]
Hi,
I am carrying on a project, which simplifying a lot, I could classify as a USB sound card.

Since my device uses its own clock for audio I have declared the following interfaces/endpoints:

- **Interface 1** containing 2 endpoint:
    - **EP 1 OUT**: Isochronous, Async, Data
    - **EP 1 IN**: Isochronous, No sync, explicit feedback
- **Interface 2** containing 1 endpoint:
    - **EP 2 IN**: Isochronous, Async, Data

This way it works quite well in Linux, the host "chases" the clock of my device even when I force it to vary (just to evaluate the goodness of my feedback implementation)

However, in the USB2 specification I read about implicit feedback:
![impl_fb](https://github.com/hathach/tinyusb/assets/31435395/544bb472-759d-42a2-a987-7d60f6a208c7)

I wanted to try this mode as well, declaring EP 2 IN as implicit feedback instead of Data, either by leaving the implicit feedback EP in as well or removing it, however in neither scenario does it work. The device is correctly enumerated, but when I select it in some SW (such as Ardour), ALSA, almost immediately, reports a generic "I/O Error"

Perhaps EP 1 OUT and EP 2 IN are not properly grouped, as is required by the specification. To group them, do I have to put them both in one interface? In this way, though, it will no longer be possible to open one of the two streams only?

@Protoxy22 and I were talking about this [here](https://github.com/hathach/tinyusb/pull/2328).
He asked me to share some files with him, which I attach in this post.
Hey, @Protoxy22! The files are here! ;-)
[usb_desc.zip](https://github.com/hathach/tinyusb/files/14310624/usb_desc.zip)

**Comment by Protoxy22** [comment]


**Reply by mastupristi** [reply]
Hi @Protoxy22 

> At first glance for the implicit feedback descriptor, I would do this instead (the opposite of what you did)
> 
> `TUD_AUDIO_DESC_STD_AS_ISO_EP(_epout, (uint8_t) (TUSB_XFER_ISOCHRONOUS | TUSB_ISO_EP_ATT_ASYNCHRONOUS | TUSB_ISO_EP_ATT_IMPLICIT_FB)`
> 
> `TUD_AUDIO_DESC_STD_AS_ISO_EP(_epin, (uint8_t) (TUSB_XFER_ISOCHRONOUS | TUSB_ISO_EP_ATT_ASYNCHRONOUS | TUSB_ISO_EP_ATT_DATA)`
> 
> It's sad but the host (PC) will always be the master of all transactions, you can't tell him (from device) to reduce or increase the epout datarate.

It is not clear to me why you're saying this. USB2 specifications say exactly the opposite. For this purpose, they specified the feedback mechanism (explicit or implicit). And in the "USB Device Class Definition for Audio Devices" it is clearly written that for an asynchronous sink (EP OUT) you need explicit feedback
![UAC2_000](https://github.com/hathach/tinyusb/assets/31435395/7644afc3-f572-4771-af93-d2bb7d19570e)

Also, for asynchronous sources (EP IN) the host should adjust the data rate following what comes to it from the device
![UAC_async_source](https://github.com/hathach/tinyusb/assets/31435395/0ddca709-eb93-4f7e-a0b5-ce961b89c7f5)

Finally, I implemented explicit feedback first. And I did some targeted testing to assess when the host "adjusts" to my rate: I have a command that allows me to vary the clock of the SAIs even by a lot (+/- 2000ppm).
In this way I could verify that the host really "chases" my clock.

That said and given what was written in 5.12.4.3, I wanted to try it without explicit feedback but only with implicit feedback. 

I will repost the important detail in my opinion:
![UAC2_impl_fd_det](https://github.com/hathach/tinyusb/assets/31435395/91cbb21d-a314-46b4-9afe-bb4f13547dd7)

My case is exactly this.

If I swap the data EP with the Implicit fb EP, I can't figure out how it would work

> I think in implicit feedback the host will send variable audio packet size, so you have to audio_write() the same data size of the latest received packet otherwise there will be an error on your audio driver

The host adjusts its rate based on feedback (implicit or explicit). This is done by varying the number of slots it sends in a unit of time.
For example, at theoretical 48kHz I have 48000 slots every second, and on USB HS I have 8000 microframes. Consequently if each microframe carries 6 slots, you collect the desired 48000 slots every second.
Suppose though that my device clock is slower and I have an actual frequency 47980Hz. In one second the host generates 20 microframes containing 5 slots, while all the other microframes have 6 slots.

This is well described in 5.12.4.2

If you are interested, I found that when I declare EP 2 implicit feedback, it seems that the interrupt (relative to data) is no longer triggered.
I opened a thread here: https://github.com/hathach/tinyusb/discussions/2476

best regards
Max
